### PR TITLE
Transient failure retries

### PR DESCRIPTION
**Pull request by tvrprasad**
@arthurschreiber @chdh This adds retries for transient failures that all Microsoft drivers do to improve reliability. This could address some of the connection failures issues that have been reported over time.

### CODE REVIEW

**Comment by tvrprasad** [comment]
Some tests failing on Linux for NodeJS 4. Error suggests issue with machine state than code:
> Failed to connect to localhost:1433 in 15000ms

Is there some way to retrigger the test without submitting a dummy change?

**Comment by tvrprasad** [comment]
I made a dummy PR (https://github.com/tediousjs/tedious/pull/575) with the same changes which triggered a re-run and tests pass now.

File: src/transient-error-lookup.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+class TransientErrorLookup {
```
Comment (chdh) [inline src/transient-error-lookup.js]: Why is the isTransientError() function within a class?

File: src/transient-error-lookup.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  isTransientError(error) {
+    // This list of transient errors comes from Microsoft implementation of SqlClient:
+    //  - https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlInternalConnectionTds.cs#L115
+    const transientErrors = [4060, 10928, 10929, 40197, 40501, 40613];
+
+    for (let j = 0; j < transientErrors.length; j++) {
```
Comment (chdh) [inline src/transient-error-lookup.js]: Why not use Array.indexOf()? It's available in Node 4.8.3.

File: src/connection.js [hunk 3]
```diff
@@ -331,6 +361,12 @@ class Connection extends EventEmitter {
           return this.request.error.lineNumber = token.lineNumber;
         }
       } else {
+        const isLoginErrorTransient = this.transientErrorLookup.isTransientError(token.number);
+        if (isLoginErrorTransient && this.curTransientRetryCount !== this.config.options.maxRetriesOnTransientErrors) {
+          this.debug.log('Initiating retry on transient error = ', token.number);
+          return this.transitionTo(this.STATE.TRANSIENT_FAILURE_RETRY);
```
Comment (chdh) [inline src/connection.js]: The "return" statements with a value are an artefact of the CoffeeScript to JavaScript conversion and should not be used whereit makes no sense.

File: src/connection.js [hunk 4]
```diff
@@ -331,6 +361,12 @@ class Connection extends EventEmitter {
           return this.request.error.lineNumber = token.lineNumber;
         }
       } else {
+        const isLoginErrorTransient = this.transientErrorLookup.isTransientError(token.number);
+        if (isLoginErrorTransient && this.curTransientRetryCount !== this.config.options.maxRetriesOnTransientErrors) {
```
Comment (chdh) [inline src/connection.js]: Better use `<` than `!==` for comparing a retry count.

File: src/connection.js [hunk 5]
```diff
@@ -663,6 +717,12 @@ class Connection extends EventEmitter {
     if (this.state === this.STATE.REROUTING) {
       this.debug.log('Rerouting to ' + this.routingData.server + ':' + this.routingData.port);
       return this.dispatchEvent('reconnect');
+    } else if (this.state === this.STATE.TRANSIENT_FAILURE_RETRY) {
+      const server = this.routingData ? this.routingData.server : this.server;
+      const port = this.routingData ? this.routingData.port : this.config.options.port;
+      this.debug.log('Retry after transient failure connecting to ' + server + ':' + port);
+
+      return this.dispatchEvent('retry');
```
Comment (chdh) [inline src/connection.js]: Why not just call `createRetryTimer()` here? `dispatchEvent()` is executed synchronously.

File: test/unit/transient-error-lookup-test.js [hunk 6]
```diff
@@ -0,0 +1,20 @@
+const TransientErrorLookup = require('../../src/transient-error-lookup').TransientErrorLookup;
+
```
Comment (chdh) [inline test/unit/transient-error-lookup-test.js]: This test case seems very superfluous to me for such a simple function.

File: test/integration/connection-retry-test.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    this.sinon.restore();
+    done();
+  },
+
+  'retry specified number of times on transient errors': function(test) {
+    const config = getConfig();
+    test.expect(config.options.maxRetriesOnTransientErrors + 1);
+
+    this.sinon.stub(TransientErrorLookup.prototype, 'isTransientError', () => {
+      return true;
```
Comment (chdh) [inline test/integration/connection-retry-test.js]: Why not test for the error code for "invalid password" here?

File: src/connection.js (hunk 3 above)
Comment (tvrprasad) [inline src/connection.js]: Done.
Added a clean-up item to remove these across the board - https://github.com/tediousjs/tedious/issues/576

File: src/connection.js (hunk 5 above)
Comment (tvrprasad) [inline src/connection.js]: I think that'd make it harder to follow the logic. Now it's easy to see what's happening by looking at code under <code>TRANSIENT_FAILURE_RETRY</code> and back track from there.

File: src/transient-error-lookup.js (hunk 1 above)
Comment (tvrprasad) [inline src/transient-error-lookup.js]: This makes it easier and cleaner to stub it out in the tests. Add a comment to clarify that.

File: src/transient-error-lookup.js (hunk 2 above)
Comment (tvrprasad) [inline src/transient-error-lookup.js]: Done.

File: test/integration/connection-retry-test.js (hunk 7 above)
Comment (tvrprasad) [inline test/integration/connection-retry-test.js]: Done.

File: test/unit/transient-error-lookup-test.js (hunk 6 above)
Comment (tvrprasad) [inline test/unit/transient-error-lookup-test.js]: Makes it even more silly now that I changed the code to indexOf :-)
This is just an assertion to make sure any additions or deletion from list of transient errors is intentional.

Comment (tvrprasad) [inline test/unit/transient-error-lookup-test.js]: btw, this test actually caught a bug when I switched to Array.indexOf per your suggestion. I had <code>return transientErrors.indexOf(error) === -1;</code> where I should have <code>return transientErrors.indexOf(error) !== -1;</code> and this test failed.

**Comment by tvrprasad** [comment]
@chdh Addressed/responded to your comments. If you don't have more comments I'll commit this. Thanks!

**Pull request merged**